### PR TITLE
Avoid adding builtin functions to `symbols.o` 

### LINKED ISSUE #118559: Windows x64 no-std build broken with errors related to compiler-builtins
Hello, the nightly update on 2023-12-02 seems to break the continuous integration (CI) on the [basm-rs](https://github.com/byeongkeunahn/basm-rs) project on Windows x64 builds (nightly-x86_64-pc-windows-msvc). The Linux x86/x64 builds and the wasm32 builds continue to run fine. The recent commits on the repo is Linux-only and not related to Windows; thus the Windows build should not be affected. It seems like the recently merged pull request [113923](https://github.com/rust-lang/rust/pull/113923) may be relevant.

I expected to see this happen: the basm-rs CI build should continue to pass.

Instead, this happened: the basm-rs CI build fails with a linker error related to compiler-builtins.

### Meta
I tried to localize the error with `cargo-bisect-rustc` on Windows 11 x64 with the msvc toolchain as follows:
```
git clone https://github.com/byeongkeunahn/basm-rs.git
cargo-bisect-rustc -vv --start=2023-12-01 --end=2023-12-03
```
which results in the following analysis report:

searched nightlies: from nightly-2023-12-01 to nightly-2023-12-03
regressed nightly: nightly-2023-12-02
searched commit range: https://github.com/rust-lang/rust/compare/87e1447aadaa2899ff6ccabe1fa669eb50fb60a1...8c2b57721728233e074db69d93517614de338055
regressed commit: https://github.com/rust-lang/rust/commit/8c2b57721728233e074db69d93517614de338055

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>

Host triple: x86_64-pc-windows-msvc
Reproduce with:
```bash
cargo bisect-rustc --start=2023-12-01 --end=2023-12-03
```
</details>

The actual error message is about unresolved external symbols.
```
  = note: symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem4bcmp17hdd43ed361972df2aE
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem6memcmp17h4000bdfb54398fd7E
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem6memcpy17h76ab47a482abb17aE
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem6memset17hddee021ef3511addE
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem6strlen17h8221a76238dd80f9E
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem7memmove17h9a354b96e96d4c1eE
          ..(omitted)\target-bisector-ci-8c2b57721728233e074db69d93517614de338055-x86_64-pc-windows-msvc\x86_64-pc-windows-msvc\debug\deps\basm.exe : fatal error LNK1120: 6 unresolved externals
```

Thanks in advance!

### PR DESCRIPTION
**Pull request by dianqk**
We found performance regressions in #113923. The problem seems to be that `--gc-sections` does not remove these symbols. I tested that lld removes these symbols, but ld and gold do not.

I found that `used` adds symbols to `symbols.o` at https://github.com/rust-lang/rust/blob/3e202ead604be31f4c1a5798a296953d3159da7e/compiler/rustc_codegen_ssa/src/back/linker.rs#L1786-L1791.
The PR removes builtin functions.

Note that under LTO, ld still preserves these symbols. (lld will still remove them.)

The first commit also fixes #118559. But I think the second commit also makes sense.

### CODE REVIEW

**Comment by DianQK** [comment]
r? @bjorn3

cc @nnethercote 

We probably can perform a performance test.

**Comment by bjorn3** [comment]
`symbols.o` is necessary to ensure cyclic dependencies work correctly by forcing that the object file that defines them is pulled in and that the linker will consider this object file as a way to resolve uses of this symbol. This applies not just to linker-used symbols. If not including all these in symbols.o, we need some other way to make the linker behave in this way.

**Comment by DianQK** [comment]
> `symbols.o` is necessary to ensure cyclic dependencies work correctly by forcing that the object file that defines them is pulled in and that the linker will consider this object file as a way to resolve uses of this symbol. This applies not just to linker-used symbols. If not including all these in symbols.o, we need some other way to make the linker behave in this way.

I moved `#[used(compiler)]` back to `used`. This should keep the original behavior.

File: compiler/rustc_middle/src/middle/exported_symbols.rs [hunk 1]
```diff
@@ -36,6 +36,7 @@ pub struct SymbolExportInfo {
     pub level: SymbolExportLevel,
     pub kind: SymbolExportKind,
     pub used: bool,
+    pub used_compiler: bool,
```
Comment (RalfJung) [inline compiler/rustc_middle/src/middle/exported_symbols.rs:39]: Please add a doc comment explaining why there are two `used` fields and what they each do.

File: compiler/rustc_codegen_ssa/src/back/symbol_export.rs [hunk 2]
```diff
@@ -110,7 +110,8 @@ fn reachable_non_generics_provider(tcx: TyCtxt<'_>, _: LocalCrate) -> DefIdMap<S
             // We have to preserve the symbols of the built-in functions during LTO.
             let is_builtin_fn = is_compiler_builtins
                 && symbol_export_level(tcx, def_id.to_def_id())
-                    .is_below_threshold(SymbolExportLevel::C);
+                    .is_below_threshold(SymbolExportLevel::C)
+                && name.contains("compiler_builtins");
```
Comment (bjorn3) [inline compiler/rustc_codegen_ssa/src/back/symbol_export.rs:114]: I don't think we should check the name here. The compiler-builtins crate may have a different name. Maybe check for `#[no_mangle]` instead?

**Comment by nnethercote** [comment]
It would be good to do a perf run on CI before this is merged, to confirm the expected perf effects.

File: compiler/rustc_codegen_ssa/src/back/symbol_export.rs (hunk 2 above)
Comment (DianQK) [reply compiler/rustc_codegen_ssa/src/back/symbol_export.rs]: Yeah, it's better.

**Comment by DianQK** [comment]
> It would be good to do a perf run on CI before this is merged, to confirm the expected perf effects.

Could you do this for me? I don't have perf permissions.

**Comment by saethlin** [comment]
@DianQK You can always ask for a perf run https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/perf.20run which will probably be faster than hoping someone sees a GitHub comment. I mean, maybe it won't for you, since your work attracts so many eyes. But in general, the Zulip topic is pretty quck.

**Comment by DianQK** [comment]
> @DianQK You can always ask for a perf run https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/perf.20run which will probably be faster than hoping someone sees a GitHub comment. I mean, maybe it won't for you, since your work attracts so many eyes. But in general, the Zulip topic is pretty quck.

Thanks. I saved this stream. :)

**Pull request merged**